File: FacadeDemo/FacadeDemo.java
import java.util.List;
import java.text.MessageFormat;
import java.util.ArrayList;

class Account {
    int accountNumber;
    String holdersName;
    double balance;

    public Account(int accountNumber, String holdersName, double balance) {
        this.accountNumber = accountNumber;
        this.holdersName = holdersName;
        this.balance = balance;
    }
}

class Bank {
    List<Account> accounts = new ArrayList<>();

    public Bank(Account account) {
        accounts.add(new Account(101, "Mukesh Ambani", 100_000_000));
        accounts.add(new Account(102, "Gautam Adani", 200_000_000));
        accounts.add(new Account(103, "Anil Ambani", 100_000));
    }

    public Account geAccount(String name) {
        String search_name = name.trim().toLowerCase();
        Account found = accounts
                .stream()
                .filter(account -> account.holdersName.trim().toLowerCase().equals(search_name))
                .findFirst()
                .orElse(null);

        return found;
    }
}

class Rating {
    String name;
    int rating;

    public Rating(String name, int rating) {
        this.name = name;
        this.rating = rating;
    }
}

class CreditRating {
    List<Rating> ratings = new ArrayList<>();

    public CreditRating() {
        ratings.add(new Rating("Mukesh Ambani", 900));
        ratings.add(new Rating("Gautam Adani", 750));
        ratings.add(new Rating("Anil Ambani", 250));
    }

    public boolean eligibleForLoan(String name) {
        String search_name = name.trim();
        Rating person = ratings
                .stream()
                .filter(rating -> rating.name.trim().equalsIgnoreCase(search_name))
                .findFirst()
                .orElse(null);
        if (null != person)
            return person.rating >= 650;
        return false;
    }
}

class LoanHistory {
    String name;
    int loanAmount;
    String paid;

    public LoanHistory(String name, int loanAmount, String paid) {
        this.name = name;
        this.loanAmount = loanAmount;
        this.paid = paid;
    }
}

class CreditHistory {
    List<LoanHistory> history = new ArrayList<>();

    public CreditHistory() {
        history.add(new LoanHistory("Mukesh Ambani", 100_000, "paid"));
        history.add(new LoanHistory("Mukesh Ambani", 500_000, "ongoing"));
        history.add(new LoanHistory("Gautam Ambani", 10_000_000, "ongoing"));
        history.add(new LoanHistory("Gautam Ambani", 15_000_000, "ongoing"));
        history.add(new LoanHistory("Gautam Ambani", 2_000_000, "paid"));
        history.add(new LoanHistory("Anil Ambani", 5_000_000, "paid"));
        history.add(new LoanHistory("Anil Ambani", 10_000_000, "defaulted"));
    }

    public boolean hasDefaulted(String name) {
        String search_name = name.trim();
        LoanHistory result = history
                .stream()
                .filter(his -> his.name.trim().equalsIgnoreCase(search_name)
                        && his.paid.equalsIgnoreCase("defaulted"))
                .findFirst()
                .orElse(null);
        return null != result;
    }
}

public class FacadeDemo {
    public static void main(String[] args) {
        String name = "Mukesh Ambani";
        double loanAmount = 100_000;

        Bank bank = new Bank(null);
        CreditRating rating = new CreditRating();
        CreditHistory history = new CreditHistory();
        Account account = bank.geAccount(name);

        boolean approved = true;
        if (null == account) {
            System.err.println(MessageFormat.format("Application {0} does not have a bank account", name));
            approved = false;
        }

        if (approved && !rating.eligibleForLoan(name)) {
            System.err.println(MessageFormat.format("Application {0} is not eligible for loan", name));
            approved = false;
        }

        if (approved && history.hasDefaulted(name)) {
            System.err.println(MessageFormat.format("Application {0} has defaulted before", name));
            approved = false;
        }

        if (approved && null != account) {
            if (account.balance < loanAmount / 10) {
                System.err.println(MessageFormat.format("Application {0} does not have sufficient funds", name));
                approved = false;
            }
        }
        String appStr = approved ? " " : " not ";
        System.out.println(
                MessageFormat.format("Loan of amount {0} of {1} has{2}been approved", loanAmount, name, appStr));
    }
}

File: FacadeDemo/FacadeDemo.cs
class Account
{
    public int accountNumber;
    public string holdersName;
    public double balance;

    public Account(int accountNumber, string holdersName, double balance)
    {
        this.accountNumber = accountNumber;
        this.holdersName = holdersName;
        this.balance = balance;
    }
}

class Bank
{
    List<Account> accounts = new();

    public Bank(Account account)
    {
        accounts.Add(new Account(101, "Mukesh Ambani", 100_000_000));
        accounts.Add(new Account(102, "Gautam Adani", 200_000_000));
        accounts.Add(new Account(103, "Anil Ambani", 100_000));
    }

    public Account GetAccount(string name)
    {
        string searchName = name.Trim();
        Account found = accounts
                            .Find(acc => acc.holdersName.Trim().Equals(searchName, StringComparison.OrdinalIgnoreCase));
        return found;
    }
}

class Rating
{
    public string name;
    public int rating;

    public Rating(string name, int rating)
    {
        this.name = name;
        this.rating = rating;
    }
}

class CreditRating
{
    List<Rating> ratings = new();

    public CreditRating()
    {
        ratings.Add(new Rating("Mukesh Ambani", 900));
        ratings.Add(new Rating("Gautam Adani", 750));
        ratings.Add(new Rating("Anil Ambani", 250));
    }

    public bool EligibleForLoan(string name)
    {
        string searchName = name.Trim();
        Rating person = ratings.Find(rating => rating.name.Trim().Equals(searchName, StringComparison.OrdinalIgnoreCase));
        if (null != person)
            return person.rating >= 650;
        return false;
    }
}

class LoanHistory
{
    public string name;
    public int loanAmount;
    public string paid;

    public LoanHistory(string name, int loanAmount, string paid)
    {
        this.name = name;
        this.loanAmount = loanAmount;
        this.paid = paid;
    }
}

class CreditHistory
{
    List<LoanHistory> history = new();

    public CreditHistory()
    {
        history.Add(new LoanHistory("Mukesh Ambani", 100000, "paid"));
        history.Add(new LoanHistory("Mukesh Ambani", 500000, "ongoing"));
        history.Add(new LoanHistory("Gautam Ambani", 10000000, "ongoing"));
        history.Add(new LoanHistory("Gautam Ambani", 15000000, "ongoing"));
        history.Add(new LoanHistory("Gautam Ambani", 2000000, "paid"));
        history.Add(new LoanHistory("Anil Ambani", 5000000, "paid"));
        history.Add(new LoanHistory("Anil Ambani", 10000000, "defaulted"));
    }

    public bool HasDefaulted(string name)
    {
        string searchName = name.Trim();
        LoanHistory result = history.Find(his =>
        {
            return his.name.Trim().Equals(searchName, StringComparison.OrdinalIgnoreCase)
                   && his.paid.Trim().Equals("defaulted", StringComparison.OrdinalIgnoreCase);
        });
        return null != result;
    }
}

public class FacadeDemo
{
    static void Main(string[] args)
    {
        string name = "Anil Ambani";
        double loanAmount = 100_000;

        Bank bank = new Bank(null);
        CreditRating rating = new CreditRating();
        CreditHistory history = new CreditHistory();
        Account account = bank.GetAccount(name);
        
        bool approved = true;        
        if (null == account)
        {
            Console.Error.WriteLine($"Application {name} does not have a bank account");
            approved = false;
        }

        if (!rating.EligibleForLoan(name))
        {
            Console.Error.WriteLine($"Application {name} is not eligible for loan");
            approved = false;
        }

        if (history.HasDefaulted(name))
        {
            Console.Error.WriteLine($"Application {name} has defaulted before");
            approved = false;
        }

        if (null != account)
        {
            if (account.balance < loanAmount / 10)
            {
                Console.Error.WriteLine($"Application {name} does not have sufficient funds");
                approved = false;
            }
        }
        
        string appStr = approved ? " " : " not ";
        Console.WriteLine($"Loan of amount {loanAmount} of {name} has{appStr}been approved");
    }
}
